daemon/src/peer.rs

```diff
@@ -162,16 +163,27 @@ where
     /// Start up the internal machinery to advance the underlying protocol,
     /// react to significant events and keep auxiliary tasks running.
     ///
+    /// The function only returns when an error occurs or the `shutdown` future
+    /// becomes ready.
+    ///
+    /// The future returned by this function must be run to completion for the
+    /// daemon to shut down properly.
+    ///
     /// # Errors
     /// * Failed to accept peer connections
     /// * A subroutine panicked or was cancelled
     ///
     /// # Panics
     /// * If the subroutine is gone when the protocol network is still setting
     ///   up shop.
-    pub async fn run(self) -> Result<(), Error> {
+    pub async fn run(self, shutdown: impl Future<Output = ()> + Send) -> Result<(), Error> {
```

Comment:
It's not entirely clear to me why this needs to be a future -- arguably an `AtomicBool` or somesuch is much more convenient for the caller, as it can be flipped in a `Drop` impl without hoop jumping. Afaiu, the only problem with the previous approach was that the loop was not broken by matching on the `Done` error.

Reply:
It would be possible to change the argument to `AtomicBool`. I am right that would require us to spawn another task or thread that watches the bool and then calls `stop_accepting()` when it flips to true? The seed code was accomplishing this by spawning a thread. This is something I wanted to avoid and why I went for a future. Happy to implement this differently though.

On the other hand creating the shutdown future with a `oneshot` is not to complicated.
```rust
let (shutdown_tx, shutdown_rx) = tokio::sync::oneshot::channel();
peer.run(shutdown_rx);
// ...
shutdown_tx.send()
```

versus with `AtomicBool`
```rust
let shutdown = Arc::new(AtomicBool::new(false));
peer.run(shutdown.clone());
// ...
shutdown.store(true, Ordering::Relaxed)
```

> Afaiu, the only problem with the previous approach was that the loop was not broken by matching on the Done error.

There was another issue in the seed: When the signal was received while awaiting `.bind()` the loop would not be broken because the old `accept` was stopped but not the new one.

Reply:
Mya, I'm not objecting to performing any sort of oneshot gymnastics _internally_, but I think you're unnecessarily putting the responsibility for preventing premature dropping of futures on the caller. As I can see from the companion PR, this results in avoidable complexity.

As a refresher, there are two main reasons for the requirement to run the network loop future to completion:

1. When working with multiple runtimes (such as in tests), deadlocks occur because spawned blocking tasks frequently need to re-enter the async runtime in order to pull data from a QUIC stream. Due to tokio's decidedly braindead shutdown behaviour, this can result in the await point never resolving because tokio already dropped all non-blocking futures. Because blocking tasks cannot be cancelled, the program hangs indefinitely. To avoid this, we close the endpoint and keep driving all futures until the terminate by themselves (eg. by receiving an IO error).
2. When working with a single runtime (such as in an executable), it is still important (up to and excluding `SIGKILL`) to let the QUIC stack drain all outstanding connections. Otherwise, peers won't notice that the connection is half-closed, and need to sit out the idle period. We currently are lenient in that we allow a peer to establish more than one connection, but that may change again.

With that out of the way, I think the simplest (read: hardest to fuck up) API for this is to _return_ a callable which shuts down the stack, a la:

```rust
pub fn run(self) -> (Shutdown, impl Future<Output = Result<!, Error>)
```

Where `Shutdown` contains either some `AtomicRef`-like with the current shutdown function, or is wired up using a channel. Either way, it's trigger would be a blocking function, or a `Drop` impl at your discretion.

The UX is then roughly:

```rust
let (kill, run) = peer.run();
spawn(run);
// do stuff
drop(kill)
```

Which arguably is a lot simpler than what you have now.

Reply:
> With that out of the way, I think the simplest (read: hardest to fuck up) API for this is to _return_ a callable which shuts down the stack

Makes sense. I’ve implemented this in https://github.com/radicle-dev/radicle-link/pull/731/commits/166cd93ffabf3e69f6d0077ae4eee17afccb7756.